Update of the International Geomagnetic Reference Field from Version 12 (default) to Version 14. Updated coefficients
in the g1 parameter cofrm subroutine (provided at https://www.ncei.noaa.gov/products/international-geomagnetic-reference-field) and updated write dates.
Intended for use up to 2035, but warnings about using a future geomagnetic field are written to the output file from post-2030.
 
Author: Matthew Brown, Space Environment & Radio Engineering Research Group (SERENE), University of Birmingham.

diff -Nru tiegcm2.0.orig/src/apex.F90 tiegcm2.0/src/apex.F90
--- tiegcm2.0.orig/src/apex.F90	2025-01-23 15:38:29.698610000 +0000
+++ tiegcm2.0/src/apex.F90	2025-01-23 15:38:53.382739000 +0000
@@ -1864,19 +1864,19 @@
   integer,parameter :: n1=120, n2=195, isv=0
   integer,parameter :: &
     ncn1=19, & ! number of coefficients dimensioned n1
-    ncn2=6     ! number of coefficients dimensioned n2: increase with each IGRF update
+    ncn2=8     ! number of coefficients dimensioned n2: increase with each IGRF update
   integer,parameter :: ngh = n1*ncn1 + n2*ncn2 + 1 ! not sure why the extra +1
   real,save :: g1(n1,ncn1), g2(n2,ncn2), gh(ngh)
   real,parameter :: alt = 0.
 
-  if (date < 1900. .or. date > 2025.) then
-    write(6,"('>>> cofrm: date=',f8.2,' Date must be >= 1900 and <= 2025')") date
+  if (date < 1900. .or. date > 2035.) then
+    write(6,"('>>> cofrm: date=',f8.2,' Date must be >= 1900 and <= 2035')") date
     stop 'cofrm'
   endif
-  if (date > 2020.) then
+  if (date > 2030.) then
     write(6,"('>>> WARNING cofrm:')")
     write(6,"(/,'   This version of IGRF is intended for use up to ')")
-    write(6,"('     2020. Values for ',f9.3,' will be computed but')") date
+    write(6,"('     2030. Values for ',f9.3,' will be computed but')") date
     write(6,"('     may be of reduced accuracy.',/)")
   endif
 
@@ -2325,49 +2325,109 @@
     	 -0.53,   -0.26,   -0.79/)			  	
   
    g2(:,5) = (/                                               & ! 2015
-    -29442.0,-1501.0, 4797.1,-2445.1, 3012.9,-2845.6, 1676.7, &  
-      -641.9, 1350.7,-2352.3, -115.3, 1225.6,  244.9,  582.0, &   
-      -538.4,  907.6,  813.7,  283.3,  120.4, -188.7, -334.9, &   
-       180.9,	70.4, -329.5, -232.6,  360.1,	47.3,  192.4, &   
-       197.0, -140.9, -119.3, -157.5,	16.0,	 4.1,  100.2, &   
-    	70.0,	67.7,  -20.8,	72.7,	33.2, -129.9,	58.9, &   
-       -28.9,  -66.7,	13.2,	 7.3,  -70.9,	62.6,	81.6, &   
-       -76.1,  -54.1,	-6.8,  -19.5,	51.8,	 5.7,	15.0, &   
-    	24.4,	 9.4,	 3.4,	-2.8,  -27.4,	 6.8,	-2.2, &   
-    	24.2,	 8.8,	10.1,  -16.9,  -18.3,	-3.2,	13.3, &   
-       -20.6,  -14.6,	13.4,	16.2,	11.7,	 5.7,  -15.9, &   
-    	-9.1,	-2.0,	 2.1,	 5.4,	 8.8,  -21.6,	 3.1, &   
-    	10.8,	-3.3,	11.8,	 0.7,	-6.8,  -13.3,	-6.9, &   
-    	-0.1,	 7.8,	 8.7,	 1.0,	-9.1,	-4.0,  -10.5, &   
-    	 8.4,	-1.9,	-6.3,	 3.2,	 0.1,	-0.4,	 0.5, &   
-    	 4.6,	-0.5,	 4.4,	 1.8,	-7.9,	-0.7,	-0.6, &   
-    	 2.1,	-4.2,	 2.4,	-2.8,	-1.8,	-1.2,	-3.6, &   
-    	-8.7,	 3.1,	-1.5,	-0.1,	-2.3,	 2.0,	 2.0, &   
-    	-0.7,	-0.8,	-1.1,	 0.6,	 0.8,	-0.7,	-0.2, &   
-    	 0.2,	-2.2,	 1.7,	-1.4,	-0.2,	-2.5,	 0.4, &   
-    	-2.0,	 3.5,	-2.4,	-1.9,	-0.2,	-1.1,	 0.4, &   
-    	 0.4,	 1.2,	 1.9,	-0.8,	-2.2,	 0.9,	 0.3, &   
-    	 0.1,	 0.7,	 0.5,	-0.1,	-0.3,	 0.3,	-0.4, &   
-    	 0.2,	 0.2,	-0.9,	-0.9,	-0.1,	 0.0,	 0.7, &   
-    	 0.0,	-0.9,	-0.9,	 0.4,	 0.4,	 0.5,	 1.6, &   
-    	-0.5,	-0.5,	 1.0,	-1.2,	-0.2,	-0.1,	 0.8, &   
-    	 0.4,	-0.1,	-0.1,	 0.3,	 0.4,	 0.1,	 0.5, &   
-    	 0.5,	-0.3,	-0.4,	-0.4,	-0.3,	-0.8/)	         
+    -29441.46,-1501.77,  4795.99, -2445.88,  3012.20, -2845.41,  1676.35, &
+     -642.17,  1350.33, -2352.26,  -115.29,  1225.85,   245.04,   581.69, &
+     -538.70,   907.42,   813.68,   283.54,   120.49,  -188.43,  -334.85, &
+      180.95,    70.38,  -329.23,  -232.91,   360.14,    46.98,   192.35, &
+      196.98,  -140.94,  -119.14,  -157.40,    15.98,     4.30,   100.12, &
+       69.55,    67.57,   -20.61,    72.79,    33.30,  -129.85,    58.74, &
+      -28.93,   -66.64,    13.14,     7.35,   -70.85,    62.41,    81.29, &
+      -75.99,   -54.27,    -6.79,   -19.53,    51.82,     5.59,    15.07, &
+       24.45,     9.32,     3.27,    -2.88,   -27.50,     6.61,    -2.32, &
+       23.98,     8.89,    10.04,   -16.78,   -18.26,    -3.16,    13.18, &
+      -20.56,   -14.60,    13.33,    16.16,    11.76,     5.69,   -15.98, &
+       -9.10,    -2.02,     2.26,     5.33,     8.83,   -21.77,     3.02, &
+       10.76,    -3.22,    11.74,     0.67,    -6.74,   -13.20,    -6.88, &
+       -0.10,     7.79,     8.68,     1.04,    -9.06,    -3.89,   -10.54, &
+        8.44,    -2.01,    -6.26,     3.28,     0.17,    -0.40,     0.55, &
+        4.55,    -0.55,     4.40,     1.70,    -7.92,    -0.67,    -0.61, &
+        2.13,    -4.16,     2.33,    -2.85,    -1.80,    -1.12,    -3.59, &
+       -8.72,     3.00,    -1.40,     0.00,    -2.30,     2.11,     2.08, &
+       -0.60,    -0.79,    -1.05,     0.58,     0.76,    -0.70,    -0.20, &
+        0.14,    -2.12,     1.70,    -1.44,    -0.22,    -2.57,     0.44, &
+       -2.01,     3.49,    -2.34,    -2.09,    -0.16,    -1.08,     0.46, &
+        0.37,     1.23,     1.75,    -0.89,    -2.19,     0.85,     0.27, &
+        0.10,     0.72,     0.54,    -0.09,    -0.37,     0.29,    -0.43, &
+        0.23,     0.22,    -0.89,    -0.94,    -0.16,    -0.03,     0.72, &
+       -0.02,    -0.92,    -0.88,     0.42,     0.49,     0.63,     1.56, &
+       -0.42,    -0.50,     0.96,    -1.24,    -0.19,    -0.10,     0.81, &
+        0.42,    -0.13,    -0.04,     0.38,     0.48,     0.08,     0.48, &
+        0.46,    -0.30,    -0.35,    -0.43,    -0.36,    -0.71/)
+
+   g2(:,6) = (/                                               & ! 2020
+    -29403.41,-1451.37,  4653.35, -2499.78,  2981.96, -2991.72,  1676.85, &
+     -734.62,  1363.00, -2380.80,   -81.96,  1236.06,   241.80,   525.60, &
+     -542.52,   902.82,   809.47,   282.10,    86.18,  -158.50,  -309.47, &
+      199.75,    47.44,  -350.30,  -234.42,   363.26,    47.52,   187.86, &
+      208.36,  -140.73,  -121.43,  -151.16,    32.09,    13.98,    99.14, &
+       65.97,    65.56,   -19.22,    72.96,    25.02,  -121.57,    52.76, &
+      -36.06,   -64.40,    13.60,     8.96,   -64.80,    68.04,    80.54, &
+      -76.63,   -51.50,    -8.23,   -16.85,    56.45,     2.36,    15.80, &
+       23.56,     6.30,    -2.19,    -7.21,   -27.19,     9.77,    -1.90, &
+       23.66,     9.74,     8.43,   -17.49,   -15.23,    -0.49,    12.83, &
+      -21.07,   -11.76,    15.28,    14.94,    13.65,     3.62,   -16.59, &
+       -6.90,    -0.34,     2.90,     5.03,     8.36,   -23.44,     2.84, &
+       11.04,    -1.48,     9.86,    -1.14,    -5.13,   -13.22,    -6.20, &
+        1.08,     7.79,     8.82,     0.40,    -9.23,    -1.44,   -11.86, &
+        9.60,    -1.84,    -6.25,     3.38,    -0.11,    -0.18,     1.66, &
+        3.50,    -0.86,     4.86,     0.65,    -8.62,    -0.88,    -0.11, &
+        1.88,    -4.26,     1.44,    -3.43,    -2.38,    -0.10,    -3.84, &
+       -8.84,     2.96,    -1.36,    -0.02,    -2.51,     2.50,     2.31, &
+       -0.55,    -0.85,    -0.39,     0.28,     0.62,    -0.66,    -0.21, &
+       -0.07,    -1.66,     1.44,    -1.60,    -0.59,    -2.98,     0.18, &
+       -1.97,     3.09,    -2.51,    -2.00,    -0.13,    -1.15,     0.43, &
+        0.52,     1.28,     1.37,    -1.14,    -1.81,     0.71,     0.08, &
+        0.31,     0.71,     0.49,    -0.15,    -0.26,     0.55,    -0.47, &
+        0.16,     0.09,    -0.93,    -1.13,    -0.04,    -0.33,     0.52, &
+        0.08,    -0.93,    -0.88,     0.53,     0.64,     0.72,     1.40, &
+       -0.30,    -0.38,     0.75,    -1.31,    -0.01,    -0.09,     0.76, &
+        0.29,    -0.05,    -0.11,     0.37,     0.47,     0.13,     0.54, &
+        0.45,    -0.41,    -0.46,    -0.36,    -0.40,    -0.60/)
+
+   g2(:,7) = (/                                               & ! 2025
+    -29350.0,  -1410.3,   4545.5,  -2556.2,   2950.9,  -3133.6,   1648.7, &
+      -814.2,   1360.9,  -2404.2,    -56.9,   1243.8,    237.6,    453.4, &
+      -549.6,    894.7,    799.6,    278.6,     55.8,   -134.0,   -281.1, &
+       212.0,     12.0,   -375.4,   -232.9,    369.0,     45.3,    187.2, &
+       220.0,   -138.7,   -122.9,   -141.9,     42.9,     20.9,    106.2, &
+        64.3,     63.8,    -18.4,     76.7,     16.8,   -115.7,     48.9, &
+       -40.9,    -59.8,     14.9,     10.9,    -60.8,     72.8,     79.6, &
+       -76.9,    -48.9,     -8.8,    -14.4,     59.3,     -1.0,     15.8, &
+        23.5,      2.5,     -7.4,    -11.2,    -25.1,     14.3,     -2.2, &
+        23.1,     10.9,      7.2,    -17.5,    -12.6,      2.0,     11.5, &
+       -21.8,     -9.7,     16.9,     12.7,     14.9,      0.7,    -16.8, &
+        -5.2,      1.0,      3.9,      4.7,      8.0,    -24.8,      3.0, &
+        12.1,     -0.2,      8.3,     -2.5,     -3.4,    -13.1,     -5.3, &
+         2.4,      7.2,      8.6,     -0.6,     -8.7,      0.8,    -12.8, &
+         9.8,     -1.3,     -6.4,      3.3,      0.2,      0.1,      2.0, &
+         2.5,     -1.0,      5.4,     -0.5,     -9.0,     -0.9,      0.4, &
+         1.5,     -4.2,      0.9,     -3.8,     -2.6,      0.9,     -3.9, &
+        -9.0,      3.0,     -1.4,      0.0,     -2.5,      2.8,      2.4, &
+        -0.6,     -0.6,      0.1,      0.0,      0.5,     -0.6,     -0.3, &
+        -0.1,     -1.2,      1.1,     -1.7,     -1.0,     -2.9,     -0.1, &
+        -1.8,      2.6,     -2.3,     -2.0,     -0.1,     -1.2,      0.4, &
+         0.6,      1.2,      1.0,     -1.2,     -1.5,      0.6,      0.0, &
+         0.5,      0.6,      0.5,     -0.2,     -0.1,      0.8,     -0.5, &
+         0.1,     -0.2,     -0.9,     -1.2,      0.1,     -0.7,      0.2, &
+         0.2,     -0.9,     -0.9,      0.6,      0.7,      0.7,      1.2, &
+        -0.2,     -0.3,      0.5,     -1.3,      0.1,     -0.1,      0.7, &
+         0.2,      0.0,     -0.2,      0.3,      0.5,      0.2,      0.6, &
+         0.4,     -0.6,     -0.5,     -0.3,     -0.4,     -0.5/)
 !
-    g2(1:80,6) = (/                                           & ! 2017
-       10.3,   18.1,  -26.6,   -8.7,   -3.3,  -27.4,    2.1,  &  
-      -14.1,	3.4,   -5.5,	8.2,   -0.7,   -0.4,  -10.1,  &   
-     	1.8,   -0.7,	0.2,   -1.3,   -9.1,	5.3,	4.1,  &   
-     	2.9,   -4.3,   -5.2,   -0.2,	0.5,	0.6,   -1.3,  &   
-     	1.7,   -0.1,   -1.2,	1.4,	3.4,	3.9,	0.0,  &   
-       -0.3,   -0.1,	0.0,   -0.7,   -2.1,	2.1,   -0.7,  &   
-       -1.2,	0.2,	0.3,	0.9,	1.6,	1.0,	0.3,  &   
-       -0.2,	0.8,   -0.5,	0.4,	1.3,   -0.2,	0.1,  &   
-       -0.3,   -0.6,   -0.6,   -0.8,	0.1,	0.2,   -0.2,  &   
-     	0.2,	0.0,   -0.3,   -0.6,	0.3,	0.5,	0.1,  &   
-       -0.2,	0.5,	0.4,   -0.2,	0.1,   -0.3,   -0.4,  &   
-     	0.3,	0.3,	0.0/)
-    g2(81:n2,6) = 0.0
+    g2(1:80,8) = (/                                           & ! 2027
+       12.6,   10.0,  -21.5,  -11.2,   -5.3,  -27.3,   -8.3,  &  
+      -11.1,   -1.5,   -4.4,    3.8,    0.4,   -0.2,  -15.6,  &   
+       -3.9,   -1.7,   -2.3,   -1.3,   -5.8,    4.1,    5.4,  &   
+        1.6,   -6.8,   -4.1,    0.6,    1.3,   -0.5,    0.0,  &   
+        2.1,    0.7,    0.5,    2.3,    1.7,    1.0,    1.9,  &   
+       -0.2,   -0.3,    0.3,    0.8,   -1.6,    1.2,   -0.4,  &   
+       -0.8,    0.8,    0.4,    0.7,    0.9,    0.9,   -0.1,  &   
+       -0.1,    0.6,   -0.1,    0.5,    0.5,   -0.7,   -0.1,  &   
+        0.0,   -0.8,   -0.9,   -0.8,    0.5,    0.9,   -0.3,  &   
+       -0.1,    0.2,   -0.3,    0.0,    0.4,    0.4,   -0.3,  &   
+       -0.1,    0.4,    0.3,   -0.5,    0.1,   -0.6,    0.0,  &   
+        0.3,    0.3,    0.2/)
+    g2(81:n2,8) = 0.0
 !
 ! Set gh from g1,g2:
 !
@@ -2383,7 +2443,7 @@
   enddo
   gh(ngh) = 0. ! not sure why gh is dimensioned with the extra element, so set it to 0. 
   
-  if (date < 2015.) then
+  if (date < 2025.) then
     t   = 0.2*(date - 1900.)
     ll  = t
     one = ll
@@ -2405,8 +2465,8 @@
       tc = -0.2
       t = 0.2
     endif
-  else ! date >= 2015
-    t     = date - 2015.0
+  else ! date >= 2025
+    t     = date - 2025.0
     tc    = 1.0
     if (isv.eq.1) then
       t = 1.0
@@ -2416,7 +2476,7 @@
     nmx   = 13
     nc    = nmx*(nmx+2)
     kmx   = (nmx+1)*(nmx+2)/2
-  endif ! date < 2015
+  endif ! date < 2025
   r = alt
   l = 1
   m = 1
